src/cryptonote_core/quorum_cop.cpp

```diff
@@ -0,0 +1,221 @@
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "service_node_deregister.h"
+#include "service_node_list.h"
+#include "cryptonote_config.h"
+#include "cryptonote_core.h"
+#include "quorum_cop.h"
+
+namespace service_nodes
+{
+  quorum_cop::quorum_cop(cryptonote::core& core, service_nodes::service_node_list& service_node_list)
+    : m_core(core), m_service_node_list(service_node_list), m_last_height(0)
+  {
+  }
+
+  void quorum_cop::blockchain_detached(uint64_t height)
+  {
+    uint64_t delta_height = std::abs(static_cast<int64_t>(m_core.get_current_blockchain_height() - height));
+    if (delta_height > REORG_SAFETY_BUFFER_IN_BLOCKS)
+    {
+      LOG_ERROR("The blockchain was detached with a change in height: " << delta_height <<
+                ", which is greater than the recommended REORG_SAFETY_BUFFER_IN_BLOCKS: " << REORG_SAFETY_BUFFER_IN_BLOCKS);
+    }
+
+    m_last_height = (height < REORG_SAFETY_BUFFER_IN_BLOCKS) ? height : height - REORG_SAFETY_BUFFER_IN_BLOCKS;
+  }
+
+  static bool participates_in_quorum(const std::vector<crypto::public_key>& quorum, crypto::public_key const &my_key, size_t *index_in_quorum = nullptr)
+  {
+    size_t index = 0;
+    for (crypto::public_key const &quorum_key : quorum)
+    {
+      if (quorum_key == my_key) break;
+      ++index;
+    }
+
+    if (index_in_quorum)
+      *index_in_quorum = index;
+
+    bool result = index >= quorum.size();
+    return result;
+  }
+
+  void quorum_cop::block_added(const cryptonote::block& block, const std::vector<cryptonote::transaction>& txs)
+  {
+    crypto::public_key my_pubkey;
+    crypto::secret_key my_seckey;
+    if (!m_core.get_service_node_keys(my_pubkey, my_seckey))
+      return;
+
+    uint64_t const height        = cryptonote::get_block_height(block);
+    uint64_t const latest_height = m_core.get_current_blockchain_height();
+
+    if (latest_height < loki::service_node_deregister::VOTE_LIFETIME_BY_HEIGHT)
+      return;
+
+    uint64_t const execute_justice_from_height = latest_height - loki::service_node_deregister::VOTE_LIFETIME_BY_HEIGHT;
+    if (height < execute_justice_from_height)
+      return;
+
+    if (height == latest_height)
```

Comment:
We don't need this branch. The uptime proofs are submitted every hour by the daemon when it is idle, not just when it is in the list of nodes to test.